applications/GeoMechanicsApplication/custom_strategies/strategies/residualbased_newton_raphson_strategy_linear_elastic_dynamic.hpp

```diff
@@ -0,0 +1,503 @@
+     * @param pNewLinearSolver The linear solver employed
+     * @param pNewConvergenceCriteria The convergence criteria employed
+     * @param pNewBuilderAndSolver The builder and solver
+     * @param rParameters The parameters for the strategy
+     * @param Beta The Newmark Beta parameter
+     * @param Gamma The Newmark Gamma parameter
+     * @param MaxIterations The maximum number of non-linear iterations to be considered when solving the problem
+     * @param CalculateReactions The flag for the reaction calculation
+     * @param MoveMeshFlag The flag that allows to move the mesh
+     */
+    explicit GeoMechanicNewtonRaphsonStrategyLinearElasticDynamic(
+        ModelPart&                                 rModelPart,
+        typename TSchemeType::Pointer              pScheme,
+        typename TLinearSolver::Pointer            pNewLinearSolver,
+        typename TConvergenceCriteriaType::Pointer pNewConvergenceCriteria,
+        typename TBuilderAndSolverType::Pointer    pNewBuilderAndSolver,
+        Parameters&                                rParameters,
+        int                                        MaxIterations      = 30,
+        bool                                       CalculateReactions = false,
+        bool                                       MoveMeshFlag       = false)
+        : ResidualBasedNewtonRaphsonStrategy<TSparseSpace, TDenseSpace, TLinearSolver>(
+              rModelPart, pScheme, pNewConvergenceCriteria, pNewBuilderAndSolver, MaxIterations, CalculateReactions, false, MoveMeshFlag)
+    {
+        // new constructor
+    }
+
+    /**
+     * @brief Initialization of member variables and prior operations
+     */
+    void Initialize() override
+    {
+        KRATOS_TRY;
+
+        BaseType::Initialize();
+
+        // Note that FindNeighbourElementsOfConditionsProcess and DeactivateConditionsOnInactiveElements are required to be perfomed before initializing the System and State
+        // this means that these operations are done twice in the GeomechanicsSolver in python
+        auto find_neighbour_elements_of_conditions_process =
+            FindNeighbourElementsOfConditionsProcess(BaseType::GetModelPart());
+        find_neighbour_elements_of_conditions_process.Execute();
+
+        auto deactivate_conditions_on_inactive_elements_process =
+            DeactivateConditionsOnInactiveElements(BaseType::GetModelPart());
+        deactivate_conditions_on_inactive_elements_process.Execute();
+
+        if (!BaseType::mStiffnessMatrixIsBuilt)
+            // initialize the system matrices and the initial second derivative
+            this->InititalizeSystemAndState();
+
+        KRATOS_CATCH("");
+    }
+
+    void Predict() override
+    {
+        KRATOS_TRY
+        // OPERATIONS THAT SHOULD BE DONE ONCE - internal check to avoid repetitions
+        // if the operations needed were already performed this does nothing
+        if (BaseType::mInitializeWasPerformed == false) this->Initialize();
```

Comment:
Please use `!` rather than comparing to `false`:
```suggestion
        if (!BaseType::mInitializeWasPerformed) this->Initialize();
```

I have a conceptual question about this code as well: shouldn't `Initialize` _always_ be called prior to calling `Predict`? Are there any cases known where `Predict` can be called before calling `Initialize`? If yes, can you please explain why it has been designed that way (since this piece of code will then initialize itself anyway)?

Reply:
honestly I didnt reallt look into this, this comes from the base, but indeed in our solvers, we always call predict after initializing. I think it can indeed be removed